src/solvers/strings/string_constraint_instantiation.cpp

```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+#include <util/expr_iterator.h>
 #include "string_constraint_instantiation.h"
 
+/// look for the symbol and return true if it is found
+/// \param index: an index expression
+/// \param qvar: a symbol expression
+/// \return a Boolean
+static bool contains(const exprt &index, const symbol_exprt &qvar)
+{
+  return std::find(index.depth_begin(), index.depth_end(), qvar) !=
+         index.depth_end();
+}
+
+/// Find indexes of `str` used in `expr` that contains `qvar`, for instance
+/// with arguments ``(str[k+1]=='a')``, `str`, and `k`, the function should
+/// return `k+1`.
+/// \param [in] expr: the expression to search
+/// \param [in] str: the string which must be indexed
+/// \param [in] qvar: the universal variable that must be in the index
+/// \return index expressions in `expr` on `str` containing `qvar`.
+static std::unordered_set<exprt, irep_hash>
+find_indexes(const exprt &expr, const exprt &str, const symbol_exprt &qvar)
+{
+  decltype(find_indexes(expr, str, qvar)) result;
+  auto index_str_containing_qvar = [&](const exprt &e) {
+    if(auto index_expr = expr_try_dynamic_cast<index_exprt>(e))
+    {
+      const auto &arr = index_expr->array();
+      const auto str_it = std::find(arr.depth_begin(), arr.depth_end(), str);
+      return str_it != arr.depth_end() && contains(index_expr->index(), qvar);
+    }
+    return false;
+  };
+
+  std::for_each(expr.depth_begin(), expr.depth_end(), [&](const exprt &e) {
+    if(index_str_containing_qvar(e))
+      result.insert(to_index_expr(e).index());
+  });
+  return result;
+}
+
+std::map<exprt, int> map_representation_of_sum(const exprt &f)
+{
+  // number of time the leaf should be added (can be negative)
+  std::map<exprt, int> elems;
+
+  std::list<std::pair<exprt, bool>> to_process;
+  to_process.emplace_back(f, true);
+
+  while(!to_process.empty())
+  {
+    exprt cur = to_process.back().first;
+    bool positive = to_process.back().second;
+    to_process.pop_back();
+    if(cur.id() == ID_plus)
+    {
+      for(const auto &op : cur.operands())
+        to_process.emplace_back(op, positive);
+    }
+    else if(cur.id() == ID_minus)
+    {
+      to_process.emplace_back(cur.op1(), !positive);
+      to_process.emplace_back(cur.op0(), positive);
+    }
+    else if(cur.id() == ID_unary_minus)
+    {
+      to_process.emplace_back(cur.op0(), !positive);
+    }
+    else
+    {
+      if(positive)
+        elems[cur] = elems[cur] + 1;
+      else
+        elems[cur] = elems[cur] - 1;
+    }
+  }
+  return elems;
+}
+
+exprt
+sum_over_map(std::map<exprt, int> &m, const typet &type, bool negated)
+{
+  exprt sum = nil_exprt();
+  mp_integer constants = 0;
+  typet index_type;
+  if(m.empty())
+    return from_integer(0, type);
+  else
+    index_type = m.begin()->first.type();
+
+  for(const auto &term : m)
+  {
+    // We should group constants together...
+    const exprt &t = term.first;
+    int second = negated ? (-term.second) : term.second;
+    if(t.id() == ID_constant)
```

Comment:
I suppose this is what the comment refers to - why isn't this already done in `map_representation_of_sum`?

Reply:
`map_representation_of_sum` just maps expressions to the number of time they appear in the sum, without doing anything special for constants, I clarified that in the example for `map_representation_of_sum`

Reply:
I understand that - but is that really the most appropriate design?

Reply:
Another possibility would be to use an exprt equal to 1 in the map and add all constants to its factor. The current implementation of `sum_over_map` handles both cases so I wouldn't change it. We could make `map_representation_of_sum` do more work but I don't think there is not a huge need for it.

Reply:
My interpretation of this code is that you are computing the coefficients of a polynomial. (I do not have a sufficient understanding of the code to tell why you need to do this, but at least that's my reading of this code.) If this interpretation is correct, then I'd consider renaming the function to make this clear. And if you're doing this, then certainly constants should be folded to become the coefficient of the zero-degree summand.

Reply:
It's actually a linear function rather than a polynomial function, and computing this representation is useful to inverse the function. I've put this in a class `linear_functiont` to make that clearer (constants are now stored in a single field).